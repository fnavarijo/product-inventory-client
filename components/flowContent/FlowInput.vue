<template>
  <ValidationProvider
    v-slot="{ errors }"
    class="flex flex-col"
    :rules="rules"
  >
    <label :for="name">{{ label }}</label>
    <div class="flex">
      <input
        class="border-2 border-gray-500 p-3 w-full"
        :name="name"
        :type="type"
        :value="value"
        @input="emitInput"
      >
      <div
        v-show="shouldDisplayError(errors)"
        class="bg-red-500 px-2 flex items-center justify-center"
      >
        <WarningIcon class="w-8 h-8 text-white" />
      </div>
    </div>
    <span
      v-show="shouldDisplayError(errors)"
      class="input-error text-red-500 font-bold mt-1"
    >
      {{ errors[0] }}
    </span>
  </ValidationProvider>
</template>

<script lang="ts">
import Vue from 'vue';
import { ValidationProvider } from 'vee-validate';

import WarningIcon from '@/assets/img/warning.svg?inline';

export default Vue.extend({
  name: 'FlowInput',
  components: {
    WarningIcon,
    ValidationProvider,
  },
  props: {
    type: {
      type: String,
      default: 'text',
    },
    name: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      required: true,
    },
    value: {
      type: String,
      required: true,
    },
    rules: {
      type: String,
      default: '',
    },
  },
  methods: {
    emitInput (event: any): void {
      this.$emit('input', event.target.value);
    },
    shouldDisplayError (errors: Array<any>): boolean {
      return errors.length > 0;
    },
  },
});
</script>
