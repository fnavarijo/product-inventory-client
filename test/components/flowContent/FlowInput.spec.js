import { extend } from 'vee-validate';
import { mount } from '@vue/test-utils';
import flushPromises from 'flush-promises';

import FlowInput from '@/components/flowContent/FlowInput.vue';

describe('FlowInput component', () => {
  beforeAll(() => {
    extend('isvalid', {
      validate: (value) => { console.log('Validating'); return value === 'isValid'; },
      message: 'Should be isValid value',
    });
  });

  it('should generate input with custom design', () => {
    const wrapper = mount(FlowInput, {
      propsData: {
        name: 'username',
        label: 'User name',
        value: '',
      },
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should handle v-model', () => {
    const MountComponent = {
      template: '<FlowInput name="name" label="name" v-model="inputValue" />',
      components: { FlowInput },
      data () {
        return { inputValue: '' };
      },
    };
    const wrapper = mount(MountComponent);

    const inputElement = wrapper.find('input');
    inputElement.setValue('Hello World');

    expect(wrapper.vm.inputValue).toBe('Hello World');
  });

  it('should display error message when input is invalid', async () => {
    // TODO: is not running the validation defined in the rule
    const wrapper = mount(FlowInput, {
      propsData: {
        name: 'inputError',
        label: 'Error input',
        value: '',
        rules: 'isvalid',
      },
    });

    wrapper.find('input').setValue('invalid');

    await flushPromises();

    expect(wrapper.find('.input-error').text()).toContain('Should be isValid value');
    expect(wrapper.html()).toMatchSnapshot();
  });
});
