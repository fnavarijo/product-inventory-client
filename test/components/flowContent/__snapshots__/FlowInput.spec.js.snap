// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowInput component should display error message when passing prop 1`] = `
<span class="flex flex-col"><label for="inputError">Error input</label> <div class="flex"><input name="inputError" type="text" class="border-2 border-gray-500 p-3 w-full"> <div class="bg-red-500 px-2 flex items-center justify-center"><svg class="w-8 h-8 text-white"></svg></div></div> <span class="input-error text-red-500 font-bold mt-1">
    Something wrong happened
  </span></span>
`;

exports[`FlowInput component should generate input with custom design 1`] = `
<span class="flex flex-col"><label for="username">User name</label> <div class="flex"><input name="username" type="text" class="border-2 border-gray-500 p-3 w-full"> <div class="bg-red-500 px-2 flex items-center justify-center" style="display: none;"><svg class="w-8 h-8 text-white"></svg></div></div> <span class="input-error text-red-500 font-bold mt-1" style="display: none;">
    
  </span></span>
`;

exports[`FlowInput component should generate input with customs styles 1`] = `
<span class="flex flex-col"><label for="username">User name</label> <div class="flex"><input name="username" type="text" class="border-2 border-gray-500 p-3 w-full"> <div class="bg-red-500 px-2 flex items-center justify-center" style="display: none;"><svg class="w-8 h-8 text-white"></svg></div></div> <span class="input-error text-red-500 font-bold mt-1" style="display: none;">
    
  </span></span>
`;
